### PR TITLE
Segfault of is.sorted on list/vector containing only NA_character_

### LINKED ISSUE #5070: Segfault merging data.tables with keyed `NA_character_` columns
### Issue

Merging two data.tables where one data.table (or both) has a keyed column containing only `NA_character_`'s produces a segfault and crashes the R session.

### Reproducible example

```{r}
library(data.table)

dt1 <- data.table(x1 = rep(letters[1:4], each = 3), x2 = NA_character_)
dt2 <- data.table(x1 = letters[1:3])
  
setkey(dt1, x2)

dt3 <- dt1[dt2, on = "x1"]
dt3[, .(x1, x2)]
```

With valgrind enabled, using the current `data.table` development version (1.14.1), the above code returns:

```
==10795== Use of uninitialised value of size 8
==10795==    at 0x4FB7910: LEVELS (in /usr/lib/R/lib/libR.so)
==10795==    by 0x101824AA: issorted (in /home/jchau/R/x86_64-pc-linux-gnu-library/4.0/data.table/libs/datatable.so)
==10795==    by 0x4F352AB: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F7540B: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F7F66F: Rf_eval (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F8148E: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F82256: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F76908: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F7F66F: Rf_eval (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F8148E: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F82256: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4FC5362: ??? (in /usr/lib/R/lib/libR.so)
==10795==  Uninitialised value was created by a heap allocation
==10795==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==10795==    by 0x4FBE353: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4FBFE81: Rf_allocVector3 (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F892F7: R_bcEncode (in /usr/lib/R/lib/libR.so)
==10795==    by 0x50215C6: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x502163F: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x502095C: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x501FCA9: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x5021A2D: R_Unserialize (in /usr/lib/R/lib/libR.so)
==10795==    by 0x5022DC9: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x5023200: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F7FBF5: Rf_eval (in /usr/lib/R/lib/libR.so)
==10795== 
==10795== Invalid read of size 2
==10795==    at 0x4FB7910: LEVELS (in /usr/lib/R/lib/libR.so)
==10795==    by 0x101824AA: issorted (in /home/jchau/R/x86_64-pc-linux-gnu-library/4.0/data.table/libs/datatable.so)
==10795==    by 0x4F352AB: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F7540B: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F7F66F: Rf_eval (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F8148E: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F82256: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F76908: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F7F66F: Rf_eval (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F8148E: ??? (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4F82256: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==10795==    by 0x4FC5362: ??? (in /usr/lib/R/lib/libR.so)
==10795==  Address 0x1000000010001 is not stack'd, malloc'd or (recently) free'd
==10795== 

 *** caught segfault ***
address (nil), cause 'unknown'

Traceback:
 1: is.sorted(jval, by = key(x))
 2: `[.data.table`(dt3, , .(x1, x2))
 3: dt3[, .(x1, x2)]
```

### Session info

```{r}
R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.14.1

loaded via a namespace (and not attached):
[1] compiler_4.0.2
```

### Note

Using an explicit `merge` does work as expected:

``` r
library(data.table)

dt1 <- data.table(x1 = rep(letters[1:4], each = 3), x2 = NA_character_)
dt2 <- data.table(x1 = letters[1:3])
  
setkey(dt1, x2)

dt3 <- merge(dt1, dt2, by = "x1")
dt3[, .(x1, x2)]

#>    x1 x2
#> 1:  a <NA>
#> 2:  a <NA>
#> 3:  a <NA>
#> 4:  b <NA>
#> 5:  b <NA>
#> 6:  b <NA>
#> 7:  c <NA>
#> 8:  c <NA>
#> 9:  c <NA>
```

### PR DESCRIPTION
**Pull request by ben-schwen**
Closes #5070

### CODE REVIEW

File: src/forder.c [hunk 1]
```diff
@@ -1290,6 +1290,7 @@ SEXP issorted(SEXP x, SEXP by)
       SEXP *xd = STRING_PTR(x);
       i = 0;
       while (i<n && xd[i]==NA_STRING) i++;
+      if (i==n) break; // xd consists only out NA_STRING #5070
       bool need = NEED2UTF8(xd[i]);
```
Comment (MichaelChirico) [inline src/forder.c:1294]: not: only of

**Review by MichaelChirico — APPROVED** [review]
awesome!

**Comment by mattdowle** [comment]
Yes: great, thanks!
`is.sorted` isn't exported though so another tweak to the news item pending ...

**Comment by mattdowle** [comment]
Thanks for adding the merge test from the issue!  I was looking at it too and it's a little odd isn't it, the crash actually happens in the `[,.(x1,x2)]` part when that calls `is.sorted`, not in the merge. It's just that the merge creates the result which is keyed by the all-NA_character_ column. Anyway, all I focus on is trying to understand and follow the tests myself, and I run the tests in last released version to make sure they do segfault there (which these do). I wasn't sure why you used `copy()` though, and I looked at the last key test and wondered about that. I ended up shortening the test block, reducing the number of objects in play so I could compare it to the example in the issue, and included checking the key rather than removing the key from both sides.

**Pull request merged**